src/bibgetter.py

```diff
@@ -42,11 +58,66 @@ def is_mathscinet_id(id: str) -> bool:
 
     A valid MathSciNet ID starts with `MR` followed by 1 to 7 digits.
     Alternatively, it starts starts with `mr:`, and then a valid MR identifier.
+    TODO: move the flexibility to the guess_mathscinet_id function, and make this
+    function accept only the canonical form (MR, then exactly seven digits).
     """
     pattern = r"^(MR|mr:MR)\d{1,7}$"
 
     return re.fullmatch(pattern, id) is not None
 
+def guess_mathscinet_id(id: str) -> str | dict[str, str] | None:
+    """
+    Guess the MathSciNet ID from the given identifier.
+    Accepts both raw MathSciNet IDs and URLs in various formats.
+    """
+    if is_mathscinet_id(id):
+        return id
+    url_pattern = r"^(?:https?://)?mathscinet\.ams\.org/mathscinet/relay-station\?mr=(\d{1,7})$"
+    match = re.match(url_pattern, id)
+    if match:
+        mr_id = f"MR{match.group(1)}"
+        if is_mathscinet_id(mr_id):
+            return mr_id
+    return None
+
+def is_doi(id: str) -> bool:
+    """
+    Check if the given string is a valid DOI identifier.
+    
+    DOIs typically start with '10.' followed by a series of numbers,
+    then a forward slash and additional characters ("suffix").
+    """
+    pattern = r"^(10\.\d{4,9}(/[-._:;()a-zA-Z0-9]+)+)|(10\.1002(/[^\s/]+)+)$"
+    return re.fullmatch(pattern, id) is not None
+
+def guess_doi(id: str) -> str | dict[str, str] | None:
+    """
+    Guess the DOI from the given identifier.
+
+    The original design principle of bibgetter was to use the canonical ID as the primary
+    record identifier for a bibliography item. This does not work for DOIs since they could
+    include parentheses and other symbols which are not allowed in record identifiers. Thus
+    for some DOIs this function returns a dictionary with 'actual_id' and 'bibtex_id' keys
+    instead of just a single string.
+
+    There seems to be no standard description of what a valid record identifier could be since
+    biber/bibtex/bibtool/etc each have their own non-identical parsers, see 
+    https://tex.stackexchange.com/a/582026 .
+    """
+    # Extract DOI from URL if needed
+    id = re.sub(
+        r"^(?:https?://)?(?:dx\.)?doi\.org/(.+)$",
+        r"\1",
+        id.rstrip("/")
+    )
+    if is_doi(id):
+        forbidden_symbols = '(){}#%\\=,"\''
```

Comment:
So it seems you are opting for the strict BibLaTeX approach? If so, can you document this explicitly?